### ISSUE TITLE
TLS handshake timeout again

### ISSUE THREAD

**Issue by amazaheri** [Azure/AKS#268]
Kubectl stopped responding starting last night 😑

Central US 1.9.2 

![8f6558dc-8cae-41ea-b2a6-eeb38a584545](https://user-images.githubusercontent.com/8312775/37800589-97a1ed36-2de0-11e8-9d9f-178511c93dcb.png)

**Comment by Zimmergren** [comment]
Hi @amazaheri, I've seen this in the past weeks a lot, normally on lower-sized nodes. So out of interest, what node size do you have configured in this cluster? Thanks

**Comment by amazaheri** [comment]
3

**Comment by amazaheri** [comment]
@slack any thoughts?

**Comment by Zimmergren** [comment]
@amazaheri sorry, I meant the size of the nodes (e.g. D2, A4, etc) :)
I've experienced that when running A4 and up, my cluster is healther than if running A2, for example. 
(And I've got more than a dozen similar experiences with size combinations and cluster failures, unfortunately).

**Comment by LohithChanda** [comment]
Hey! I've got the same issue. If there's anything to do with the size of the nodes, i am currently using B2ms.

**Comment by amazaheri** [comment]
Interesting, I scaled out my cluster from 3 to 5 using Portal and I manage to connect again!

**Comment by Zimmergren** [comment]
I share that experience @amazaheri, that sometimes they self-heal correctly and all is fine. Sometimes not so much. While understanding it's a preview, I'm evaluating putting production workloads on top of it once it hits GA and if this happens as often as I'm getting it right now, scaling in and out and hoping for the best isn't my go-to plan ;-)

**Comment by amazaheri** [comment]
yeap, it hit me again today

**Comment by giorgited** [comment]
we are running 2 clusters, one with 3x D32 nodes and other with 3x D12_v2_promo nodes. Seems like our d32 has no issues while d12 has that tls handshake error come and go. Both in central us

**Comment by sercand** [comment]
Recently I switched staging cluster from self-created with kubeadm to AKS and I saw this error every day. The cluster is the same size as before and I didn't think that it would be this much troublesome.

**Comment by necevil** [comment]
Hey guys,
@sercand @giorgited @amazaheri @Zimmergren @LohithChanda 

I am collecting info on the 'Unable to connect to the server: net/http: TLS handshake timeout' issue, costs, and workarounds in a question over on StackOverflow:
https://stackoverflow.com/questions/50726534/why-cant-kubectl-connect-to-the-azure-aks-server-managing-my-cluster-net-http

If you guys could answer a couple of these (here or above on SOF)
------------

1. Have you had the TLS issue impact your Clusters more than once?
2. How long did it take before your Cluster was back up and could be connected to (what was the outage duration)?
3. Did it 'heal' itself / resolve / come back to life on it's own or did you have to post a support ticket — and wait for MS to fix it?
4. If support, how long was that ticket posted for before the problem was solved on your Cluster?
5. Did you notice anything weird with your CPU / Network IO dropping but your disk usage jumping higher?
6. Would you be in favor of allowing AKS preview customers to create higher Severity Help Tickets (For this specific issue only) regardless of their support plan in order to achieve a more timely solution?

Feel free to respond here or (as mentioned) hit StackOverflow and I will collect the responses.

**Comment by sercand** [comment]
1. Only once
2. 5 days. It was staging cluster not a big deal.
3. It was fixed after support ticket
4. 18 hours
5. No

**Comment by LohithChanda** [comment]
1. I had this issue 3 to 4 times.
2. The cluster wasn't available until i scaled out the no. of nodes. (Accessed it with the dashboard in the mean time)
3. Didn't post any support tickets.
5. No

**Comment by Zimmergren** [comment]
1. At least four times in my central cluster the last couple of weeks. More, if I count other clusters too.
2. It varies. Sometimes it's a matter of minutes; sometimes it's a matter of hours or even days.
3. It varies. Sometimes I re-create the cluster to get quickly up an running; Other times I've posted a support ticket where they helped fix it, and sometimes after some time it has started working again.
4. Don't remember.
5. No.
6. No.

**Comment by necevil** [comment]
@LohithChanda thanks! Added the fact that Scaling up your number of nodes actually worked to my list of workaround results over on StackOverflow (https://stackoverflow.com/questions/50726534/unable-to-connect-net-http-tls-handshake-timeout-why-cant-kubectl-connect?answertab=votes#tab-top).

@Zimmergren @sercand have you guys tried scaling? 
Thanks for the response guys. Will concatenate the other stuff as hopefully a few more posts / user instances roll in so hopefully people know more about what to expect when it strikes.

**Comment by Zimmergren** [comment]
@necevil yup. It worked sometimes, where the cluster self-healed a period after scaling. It failed sometimes with the same errors. I don't consider scaling a solution to this problem, as that causes other challenges depending on how things are set up. I wouldn't trust that routine for a GA workload, that's for sure. In the current preview, it's a bit wild west (and expected), and I'm happy to blow up the cluster and create a new one when this fails continuously.

**Comment by sercand** [comment]
@necevil I tried scaling on the portal but it failed to scale.

**Comment by necevil** [comment]
@Zimmergren updated my answer regarding scaling with your comment and also indicated that the actual Act of scaling itself may fail like it did in @sercand 's case.

Can you guys think of anything else that I should add over there that I have missed thus far? 

Seems like the one piece of this that's critical is having the resources / self documentation in place to quickly re-deploy whatever clusters get impacted since the scaling may not be applicable in all scenarios (may break stuff) and may not even work to begin with.

**Comment by necevil** [comment]
Have any of you guys developed scripts / resources that you utilize to quickly re-deploy your Clusters?

I am starting to think about what it might take to automate that process but figured I would see if anyone has done any prior art /  work on the subject.

**Comment by Zimmergren** [comment]
@necevil yup, I'm re-creating a cluster in a matter of minutes and have everything back up and running without any data loss. It's pretty slick, even though it shouldn't be necessary. I cannot share the details of it right now without cleaning it up first, but am happy to help any effort to automate things if you've got a new project going.

**Comment by necevil** [comment]
@Zimmergren sounds like there could be a fun collaboration there for sure!
Before I go and create a project I would love to get a sense of the scope of work (currently recovering from Lyme disease and trying to limit WIP).

To me on my side the steps that would be required to automate my deployment are:
1. Script the creation of cluster (easy, commands already saved in my docs)
2. Wait on Cluster Creation.
2. Script Init of Helm
3. Script helm installation of Ingress nGinx
4. Wait on Public IP from Azure
5. Script helm installation of Kube-Lego
6. Script kubectl apply of production Svc, Deployments
6. Script adjustment of DNS settings (if production environment is impacted)
7. Wait on DNS change
8.  Script kubectl apply of production Ingress

It seems doable with the caveat that the swapping of DNS is sort of a pain and will almost assuredly lead to some amount of downtime for someone since that needs to swap over before we can actually receive our SSL cert from LetsEncrypt.  Not sure if that's sort of the workflow you are working with or not but figured I would check back in!

I have a build container that already contains Kubectl, Azure CLI and some other tools so probably this could all be a single script that is run with a docker command to improve portability, maybe you already have it set this way?

**Comment by sauryadas** [comment]
Are you guys still seeing this issue?

**Comment by 0t3dWCE** [comment]
Yes, got it on version 1.10.3, 3 nodes k8s cluster.
```
$ kubectl get pods
Unable to connect to the server: dial tcp: lookup zz.hcp.westus.azmk8s.io on 127.0.1.1:53: read udp 127.0.0.1:46374->127.0.1.1:53: i/o timeout
```

**Comment by necevil** [comment]
@0t3dWCE I haven't had the issue for a while but if it pops up I will try to restart the container.  My question is how did you get to the container when the control layer isn't responding?

That also looks like a different error message returned by kubectl but it seems like your logs indicate the same underlying TLS handshake error was probably the cause.

**Comment by 0t3dWCE** [comment]
>> My question is how did you get to the container when the control layer isn't responding?
1. Create VM (jump) in the same VNET as k8s nodes on azure portal.
2. Set new ssh user for k8s nodes on azure portal.
3. Ssh to jump node from step 1
4. Ssh to admin k8s node
5. Now u can inspect containers
$ docker ps -a

**Comment by jnoller** [comment]
Closing this issue as old/stale.

If this issue still comes up, please confirm you are running the latest AKS release. If you are on the latest release and the issue can be re-created outside of your specific cluster please open a new github issue.

The `TLS handshake timeout` error is a generic error thrown when kubectl can not connect correctly to the API server. This can be due to custom vnet/networking settings, port blocking, url whitelisting, an overloaded etcd/API server or another issue.
